rasterio/_io.pyx

```diff
@@ -820,13 +823,168 @@ cdef class DatasetReaderBase(DatasetBase):
         return sample_gen(self, xy, indexes)
 
 
+cdef class MemoryFileBase(object):
+    """Base for a BytesIO-like class backed by an in-memory file."""
+
```

Comment:
Currently we [can't open vsi data sources in write mode](https://github.com/mapbox/rasterio/blob/master/rasterio/_io.pyx#L888-L890). Now that we have the plumbing to write `vsimem` using the VSI IO functions, could we use apply this same technique to write to `vsis3`?